hnn_core/gui/_viz_manager.py

```diff
@@ -0,0 +1,577 @@
+        "gridspec_kw={\"height_ratios\":[1,1]}",
+        "mosaic": "01\n01",
+    },
+    "single figure": {
+        "kwargs": "",
+        "mosaic": "00\n00",
+    },
+    "2row x 2col (1:1)": {
+        "kwargs":
+        "gridspec_kw={\"height_ratios\":[1,1]}",
+        "mosaic": "01\n23",
+    },
+}
+
+
+def _idx2figname(idx):
+    return f"Figure {idx}"
+
+
+def _figname2idx(fname):
+    return int(fname.split(" ")[-1])
+
+
+def _update_ax(fig, ax, single_simulation, plot_type, plot_config):
+    """Refresh plots with simulation_data.
+
+    Parameters
+    ----------
+    fig : Figure
+        A matplotlib.figure.Figure object.
+    ax : Axes
+        matplotlib.axes.Axes
+    single_simulation : dict
+        A single simulation
+    plot_type : str
+        Type of subplots
+    plot_config : dict
+        A dict specifies the preprocessing and style of plots.
+    """
+    # Make sure that visualization does not change the original data
+    dpls_copied = copy.deepcopy(single_simulation['dpls'])
+    net_copied = copy.deepcopy(single_simulation['net'])
+
+    for dpl in dpls_copied:
+        dpl.smooth(plot_config['dipole_smooth']).scale(
+            plot_config['dipole_scaling'])
+
+    if net_copied is None:
+        print("No network data")
+        return
+
+    if plot_type == 'spikes':
+        if net_copied.cell_response:
+            net_copied.cell_response.plot_spikes_raster(ax=ax, show=False)
+
+    elif plot_type == 'input histogram':
+        # BUG: got error here, need a better way to handle exception
+        if net_copied.cell_response:
+            net_copied.cell_response.plot_spikes_hist(ax=ax, show=False)
+
+    elif plot_type == 'PSD':
+        if len(dpls_copied) > 0:
+            dpls_copied[0].plot_psd(fmin=0, fmax=50, ax=ax, show=False)
+
+    elif plot_type == 'spectogram':
+        if len(dpls_copied) > 0:
+            min_f = 10.0
+            max_f = plot_config['max_spectral_frequency']
+            step_f = 1.0
+            freqs = np.arange(min_f, max_f, step_f)
+            n_cycles = freqs / 8.
+            dpls_copied[0].plot_tfr_morlet(
+                freqs,
+                n_cycles=n_cycles,
+                colormap=plot_config['spectrogram_cm'],
+                ax=ax, show=False)
+
+    elif 'dipole@' in plot_type:
+        if len(dpls_copied) > 0:
+            plot_dipole(dpls_copied,
+                        ax=ax,
+                        layer=plot_type.split("@")[1],
+                        average=True,
+                        show=False)
+
+    elif plot_type == 'current dipole':
+        if len(dpls_copied) > 0:
+            plot_dipole(dpls_copied, ax=ax, average=True, show=False)
+
+    elif plot_type == 'layer-specific dipole':
+        if len(dpls_copied) > 0:
+            ax.remove()
+            layers = ["L2", "L5", "agg"]
+            gridspec = ax.get_subplotspec()
+            gs01 = gridspec.subgridspec(3, 1)
+            ax_l2 = fig.add_subplot(gs01[0])
```

Comment:
I think a more natural ordering for the plots would be agg->L2->5

Also can you remove the xlabel for axes 0 and 1 ? The `Time (ms)` label overlaps with the subplot

Reply:
do we need both L2, L5 as well as layer-specific?